File: slidingWindow/cppRealizations/minimumWindowSubstring.cpp
#include <string>
#include <iostream>
#include <map>
#include <climits>

using namespace std;


class Solution {
public:
    string minWindow(string s, string t) {
        if(t.size() == 0){
            return "";
        }
        map <char, int> countT;
        map <char, int> window;

        for(int i = 0; i < t.size(); i++){
            countT[t[i]]++;
        }

        int have = 0, need = countT.size();
        string res = "";
        int resLen = INT_MAX;
        
        int l = 0;
        for(int r = 0; r < s.size(); r++){
            char c = s[r];
            window[c]++;
            if(countT.find(c) != countT.end() && window[c] == countT[c]){
                have++;
            }
            while(have == need){
                // update our result
                if(r - l + 1 < resLen){
                    res = s.substr(l, r - l + 1);
                    resLen = r - l + 1;
                }
                // pop from the left of our window
                window[s[l]]--;
                if(countT.find(s[l]) != countT.end() && window[s[l]] < countT[s[l]]){
                    have--;
                }
                l++;
            }
        }
        if(resLen == INT_MAX){
            return "";
        }
        return res;
    }
};



int main(){
    string s = "ADOBECODEBANC";
    string t = "ABC";
    cout << Solution().minWindow(s, t) << endl;
    return 0;
}

File: slidingWindow/minimumWindowSubstring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == '': return ''
        countT, window = {}, {}
        for c in t:
            countT[c] = 1 + countT.get(c, 0)
        have, need = 0, len(countT)
        res, resLen = '', float('inf')
        l = 0
        for r in range(len(s)):
            c = s[r]
            window[c] = 1 + window.get(c, 0)
            if c in countT and window[c] == countT[c]:
                have += 1
            while have == need:
                # update our result
                if (r - l + 1) < resLen:
                    res = s[l:r+1]
                    resLen = (r - l + 1)
                # pop from the left of our window
                window[s[l]] -= 1
                if s[l] in countT and window[s[l]] < countT[s[l]]:
                    have -= 1
                l += 1
        return res if resLen != float('inf') else ''


        
        
if __name__ == "__main__":
    s = "ab"
    t = "b"
    print(Solution().minWindow(s, t))